Add fastpath test for mask check flag

yoavnavon opened the pull request:
Summary: Check that fastpath is taken, which type (sparsity fastpath or normal) for mask that is aligned and one that is not.

Test Plan: buck test caffe2/test:test_transformers

Differential Revision: D38259928

jbschlosser commented on test/test_transformers.py:
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
+        otherwise use fastpath with traditional tensors.
+        """
+
+        with torch.no_grad():
```
style nit: you could alternatively decorate the test with `@torch.no_grad()`

yoavnavon replied:
Nice!

jbschlosser commented on test/test_transformers.py:
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
+                with patch('torch._transformer_encoder_layer_fwd') as fastpath_mock:
+                    fastpath_mock.return_value = mock_return_value
+                    model(x, src_key_padding_mask=mask)
+                    self.assertTrue(fastpath_mock.called)
```
don't we expect this to be False for certain inputs?

yoavnavon replied:
Yes, but not in the context of the `mask_check` flag which is what we are testing right now. There are multiple cases where fastpath is not taken (like using grad, or model in training mode) but right now we are assuming those conditions are met to test just that flag.

jbschlosser replied:
Cool, I'm good with this

jbschlosser commented on test/test_transformers.py:
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
+            x = torch.Tensor([[[1, 2], [3, 4], [5, 6]]]).to(torch.float)
+
+            def _test_fastpath(model, mask, mock_return_value, nested_tensors=True):
+                with patch('torch._transformer_encoder_layer_fwd') as fastpath_mock:
```
`_transformer_encoder_layer_fwd` represents the "sparsity fastpath", correct? Maybe we should name the mock with `sparsity` to avoid ambiguity as to which fastpath is being mocked out

yoavnavon replied:
Not exactly, `_transformer_encoder_layer_fwd` represents both the "sparsity fastpath" and normal fastpath. When the method is called with nested tensors, that is the sparsity fastpath, otherwise it is normal fastpath.

jbschlosser replied:
Okay gotcha. How are we validating whether the "sparsity fastpath" or "normal fastpath" is taken? Is this based on the args passed to `_transformer_encoder_layer_fwd`?

jbschlosser commented on test/test_transformers.py:
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
+            # Mask check disabled with aligned mask results in sparsity fastpath
+            _test_fastpath(model, aligned_mask, nested_tensor_return_value, nested_tensors=True)
+
+            # TODO: Check disabled with non aligned mask should not raise any error
+            # but return unmeaningful values.
+            # _test_fastpath(model, not_aligned_mask, nested_tensor_return_value, nested_tensors=True)
```
what are we waiting on to address this TODO? Do we need to avoid a second alignment check that happens in the sparsity fastpath?

yoavnavon replied:
@mikekgfb  What do you think about this?

jbschlosser reviewed (commented):
Thanks for the tests! Just a few comments below to make sure we're validating the correct behavior

jbschlosser commented on test/test_transformers.py:
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
+                    model(x, src_key_padding_mask=mask)
+                    self.assertTrue(fastpath_mock.called)
+                    for call_args, _ in fastpath_mock.call_args_list:
+                        self.assertEqual(call_args[0].is_nested, nested_tensors)
```
Suggest adding a short comment for clarity here mentioning that receiving NT args corresponds with the "sparsity fastpath"

jbschlosser commented on test/test_transformers.py:
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
+    def test_mask_check_fastpath(self):
+        """
+        Test that fastpath is executed independently of the mask that is passed.
+        If the passed mask is left aligned, test that nested tensors are used (sparsity fastpath),
```
```suggestion
        If the passed mask is left aligned or mask_check=False, test that nested tensors are used (sparsity fastpath),
```

Commit:
Add fastpath test for mask check flag [1/2] (#82999)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82999

Check that fastpath is taken, which type (sparsity fastpath or normal) for mask that is aligned and one that is not.

Test Plan: buck test caffe2/test:test_transformers

Reviewed By: mikekgfb

Differential Revision: D38259928

fbshipit-source-id: 2bbfd2231987588eedf499385d88737ce613ecc8

jbschlosser reviewed (approved):
LGTM

jbschlosser commented:
@pytorchbot merge
